gstd/src/common/reservations.rs

```diff
@@ -0,0 +1,328 @@
+struct Reservation {
+    id: ReservationId,
+    amount: u64,
+    valid_until: u32,
+}
+
+/// Reservation manager.
+///
+/// The manager is used to control multiple gas reservations
+/// across executions. It can be used when you only care about
+/// reserved amounts and not concrete [`ReservationId`]s.
+///
+/// # Examples
+///
+/// Create gas reservations inside `init` and use them inside `handle`.
+///
+/// ```
+/// use gstd::{msg, prelude::*, Reservations};
+///
+/// static mut RESERVATIONS: Reservations = Reservations::new();
+///
+/// extern "C" fn init() {
+///     unsafe {
+///         RESERVATIONS
+///             .reserve(200_000, 50)
+///             .expect("failed to reserve gas");
+///         RESERVATIONS
+///             .reserve(100_000, 100)
+///             .expect("failed to reserve gas");
+///         RESERVATIONS
+///             .reserve(50_000, 30)
+///             .expect("failed to reserve gas");
+///     }
+/// }
+///
+/// extern "C" fn handle() {
+///     let reservation = unsafe { RESERVATIONS.try_take_reservation(100_000) };
+///     match reservation {
+///         Some(id) => {
+///             msg::send_bytes_from_reservation(
+///                 id,
+///                 msg::source(),
+///                 b"send_bytes_from_reservation",
+///                 0,
+///             )
+///             .expect("Failed to send message from reservation");
+///         }
+///         None => {
+///             msg::send_bytes(msg::source(), b"send_bytes", 0).expect("Failed to send message");
+///         }
+///     }
+/// }
+/// ```
+///
+/// # See also
+/// - [`ReservationId`](ReservationId) is used to reserve and unreserve gas
+/// for program execution later.
+#[derive(Default)]
+pub struct Reservations(Vec<Reservation>);
+
+impl Reservations {
+    /// Create a new [`Reservations`] struct.
+    pub const fn new() -> Self {
+        Reservations(Vec::new())
+    }
+
+    /// Reserve the `amount` of gas for further usage.
+    ///
+    /// `duration` is the block count within which the reservation must be used.
+    ///
+    /// # Underlying logics
+    ///
+    /// Executes for O(logN)..O(N), where N is a number of stored
+    /// reservations.
+    ///
+    /// All the reservations are kept sorted by amount in
+    /// ascending order(when amount is the same, they're sorted by time when
+    /// they expire) when inserted, so the closer inserted element to the
+    /// beginning of the underlying `Vec` the closer execution time will be
+    /// to O(N).
+    ///
+    /// Also, when the underlying `Vec` will allocate new memory
+    /// the attempt to clean expired reservations occurs to avoid memory
+    /// allocations.
+    pub fn reserve(&mut self, amount: u64, duration: u32) -> Result<()> {
+        let id = ReservationId::reserve(amount, duration)?;
+        let block_height = exec::block_height();
+        let valid_until = duration.saturating_add(block_height);
+
+        let insert_range_start = self
```

Comment:
Why not just use Btreeset?

Reply:
Using `BTreeSet` as an only storage for reservations is impossible:
1. it has incovenient API for updating (not inserting) values.
2. we query and insert types more complex than just a value.

However, we can do some research on optimising the solution. Currently I believe we have the simplest solution working close to O(logN).

Reply:
Ok, anyway to be restructured and introduces in wide trait approach with different strategies some time soon